pkg/hostagent/setup.go

```diff
@@ -301,22 +301,53 @@ func (agent *HostAgent) configureContainerIfaces(metadata *md.ContainerMetadata)
 	}
 
 	podid := metadata.Id.Namespace + "/" + metadata.Id.Pod
-	{
-		agent.indexMutex.Lock()
-		if _, ok := agent.epMetadata[podid]; !ok {
-			agent.epMetadata[podid] =
-				make(map[string]*md.ContainerMetadata)
+	agent.indexMutex.Lock()
+	if val, ok := agent.epMetadata[podid]; ok {
+		// check for any stale entry present.
+		// this is possible if we miss any register events for unconfiguring the POD
+		//  ideally epMetadata[podid] Map should contain only one entry of containerID
+		// As every pod contains one network namespace for all the containers with in the pod
+		// (i.e pause container ==> metadata.Id.ContId)
+		// if there are any stale which doesn't match the ContainerID remove it
+		// @TODO This logic needs to be revisited in case if pod contains multiple network namespaces
+		for key, v := range val {
+			if metadata.Id.ContId != key {
+				logger.Info("Stale ContainerId entry: ", key)
+				err := agent.cleanStatleMetadata(key)
+				if err == nil {
+					for _, iface := range v.Ifaces {
+						agent.deallocateIpsLocked(iface)
+					}
+					delete(agent.epMetadata[podid], key)
+				}
+			}
 		}
-		agent.epMetadata[podid][metadata.Id.ContId] = metadata
-		agent.indexMutex.Unlock()
+	} else {
+		agent.epMetadata[podid] =
+			make(map[string]*md.ContainerMetadata)
 	}
+	agent.epMetadata[podid][metadata.Id.ContId] = metadata
```

Comment:
Let's add a length check and logger.Warnf if len(epMetadata[podid] > 1).

Reply:
ok